Add quoting around index names when altering MySQL schema information

gabrielrincon opened the pull request:
This is the solution for the index name like it have a "-" in the name. The char "`" is necesary for scape the "-" char.

I have the next error:

[Doctrine\DBAL\Exception\SyntaxErrorException]                               
  An exception occurred while executing 'DROP INDEX year-month ON ajustes_man  
  uales':                                                                      
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1                                                                          
                                                                               

                                                                               
  [Doctrine\DBAL\Driver\PDOException]                                          
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1                                                                          
                                                                               

                                                                               
  [PDOException]                                                               
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1

gabrielrincon committed:
Add char `

This is the solution for the index name like it have a "-" in the name. The char `` is necesary por scape the "-" char.

I have the next error:

[Doctrine\DBAL\Exception\SyntaxErrorException]                               
  An exception occurred while executing 'DROP INDEX year-month ON ajustes_man  
  uales':                                                                      
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1                                                                          
                                                                               

                                                                               
  [Doctrine\DBAL\Driver\PDOException]                                          
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1                                                                          
                                                                               

                                                                               
  [PDOException]                                                               
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error i  
  n your SQL syntax; check the manual that corresponds to your MySQL server v  
  ersion for the right syntax to use near '-month ON ajustes_manuales' at lin  
  e 1

morozov commented:
This is non-portable and probably inconsistent across use cases. E.g. on Oracle, quoting an identifier makes it case sensitive and requires it to be quoted in every single query. As a workaround, avoid using table/column/index names which require quotation.

gabrielrincon commented:
Morsov, i am agree but the file change is 

lib/Doctrine/DBAL/Platforms/MySqlPlatform.php

For Oracle is another file.

morozov commented:
@gabrielrincon I agree, but what I'm saying is that if we want something to work in DBAL, we need to make it work the same across all supported databases. And particularly for Oracle, I don't see an easy way to implement it.

Furthermore, when letting DBAL take care of your constraints, you usually shouldn't care about their names much (unless you write some hard-core SQL with hints using constraint names). Index name is optional in index/PK create operations.

Ocramius commented on lib/Doctrine/DBAL/Platforms/MySqlPlatform.php:
```diff
@@ -654,7 +654,7 @@ protected function getPreAlterTableIndexForeignKeySQL(TableDiff $diff)
                     }
 
-                    $query = 'ALTER TABLE ' . $table . ' DROP INDEX ' . $remIndex->getName() . ', ';
+                    $query = 'ALTER TABLE ' . $table . ' DROP INDEX `' . $remIndex->getName() . '`, ';
```
The quoting of an identifier (any identifier) should be performed by using `$this->quoteStringLiteral()`, and not by hardcoding backticks

Ocramius commented on lib/Doctrine/DBAL/Platforms/MySqlPlatform.php:
```diff
@@ -973,7 +973,7 @@ public function getDropIndexSQL($index, $table=null)
         }
 
-        return 'DROP INDEX ' . $indexName . ' ON ' . $table;
+        return 'DROP INDEX `' . $indexName . '` ON ' . $table;
```
The quoting of an identifier (any identifier) should be performed by using `$this->quoteStringLiteral()`, and not by hardcoding backticks

gabrielrincon replied:
Then do i have to modify the line there:

return 'DROP INDEX $this->quoteStringLiteral()' . $indexName . '$this->quoteStringLiteral() ON ' . $table;

gabrielrincon replied:
I send the changes, Thansk very much. This is my fisrt contribution.

Ocramius reviewed (changes requested):
Tests are missing

gabrielrincon committed:
Update MySqlPlatform.php

Ocramius commented:
@gabrielrincon tests are still missing. See:

 * https://github.com/doctrine/dbal/blob/7baf44892f2cfb83cedf9b05ef409e9ac793b962/tests/Doctrine/Tests/DBAL/Platforms/MySqlPlatformTest.php
 * https://github.com/doctrine/dbal/blob/7baf44892f2cfb83cedf9b05ef409e9ac793b962/tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php

gabrielrincon committed:
UPDATING FOR TESTING PASS

gabrielrincon committed:
Updating the index name 

because if this had a "-" minus symbol generate a error.

gabrielrincon commented:
Ok, i have a error in my test, i check the test log, an this is the error:


/home/scrutinizer/build/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1032

9) Doctrine\Tests\DBAL\Platforms\MySqlPlatformTest::testGeneratesAlterTableRenameIndexUsedByForeignKeySQL
Failed asserting that Array &0 (
    0 => 'ALTER TABLE mytable DROP FOREIGN KEY fk_foo'
    1 => 'DROP INDEX &#96;idx_foo&#96; ON mytable'
    2 => 'CREATE INDEX idx_foo_renamed ON mytable (foo)'
    3 => 'ALTER TABLE mytable ADD CONSTRAINT fk_foo FOREIGN KEY (foo) REFERENCES foreign_table (id)'
) is identical to Array &0 (
    0 => 'ALTER TABLE mytable DROP FOREIGN KEY fk_foo'
    1 => 'DROP INDEX idx_foo ON mytable'
    2 => 'CREATE INDEX idx_foo_renamed ON mytable (foo)'
    3 => 'ALTER TABLE mytable ADD CONSTRAINT fk_foo FOREIGN KEY (foo) REFERENCES foreign_table (id)'
).

/home/scrutinizer/build/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:1390

Then i go to change the test, because, the correct right is the sentence new generate by apply into test.

New change line:
    1 => 'DROP INDEX &#96;idx_foo&#96; ON mytable'

The line into test:
    1 => 'DROP INDEX idx_foo ON mytable'

Remember that the error is caused because if the name of idx_foo was idx-foo the compiler crash and generate error. The solution is &#96;idx-foo&#96;

gabrielrincon commented:
The modify on the test is
https://github.com/doctrine/dbal/pull/2773/files

Ocramius commented:
@gabrielrincon please incorporate the test here, together with *additional* tests for other index names that include special characters.

gabrielrincon commented:
Hi. @Ocramius . There are the changes. https://github.com/doctrine/dbal/pull/2781
Thanks very much

Ocramius commented:
@gabrielrincon tests are still missing from this patch

Ocramius commented on lib/Doctrine/DBAL/Platforms/MySqlPlatform.php:
```diff
@@ -973,7 +973,7 @@ public function getDropIndexSQL($index, $table=null)
         }
 
-        return 'DROP INDEX ' . $indexName . ' ON ' . $table;
+        return 'DROP INDEX ' . $this->getIdentifierQuoteCharacter() . $indexName . $this->getIdentifierQuoteCharacter() . ' ON ' . $table;
```
Why do you prepend/append? Use [`$this->quoteIdentifier()`](https://github.com/gabrielrincon/dbal/blob/dca8ab3c226c3166c96452d38a4eff333e3d4b51/lib/Doctrine/DBAL/Platforms/AbstractPlatform.php#L1860)

gabrielrincon committed:
I changed for $this->quoteIdentifier() 

The changes had made in 657 and 976 lines. The old method was $this->getIdentifierQuoteCharacter()

gabrielrincon committed:
Add special quote

because index name with special character in name (needs quotation on some mysql platform.

gabrielrincon committed:
Modify the index with special char

gabrielrincon committed:
Adding quoting for special chars

gabrielrincon commented:
Hi @Ocramius . I already made all the requested changes

Ocramius commented on lib/Doctrine/DBAL/Platforms/MySqlPlatform.php:
```diff
@@ -654,7 +654,7 @@ protected function getPreAlterTableIndexForeignKeySQL(TableDiff $diff)
                     }
 
-                    $query = 'ALTER TABLE ' . $table . ' DROP INDEX ' . $remIndex->getName() . ', ';
+                    $query = 'ALTER TABLE ' . $table . ' DROP INDEX ' . $this->quoteIdentifier() . $remIndex->getName() . $this->quoteIdentifier() . ', ';
```
`quoteIdentifier` accepts a parameter

Ocramius commented on lib/Doctrine/DBAL/Platforms/MySqlPlatform.php:
```diff
@@ -973,7 +973,7 @@ public function getDropIndexSQL($index, $table=null)
         }
 
-        return 'DROP INDEX ' . $indexName . ' ON ' . $table;
+        return 'DROP INDEX ' . $this->quoteIdentifier() . $indexName . $this->quoteIdentifier() . ' ON ' . $table;
```
same - how do the tests pass? :-\

Ocramius commented on tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php:
```diff
@@ -36,7 +36,7 @@ public function testInitializesJsonTypeMapping()
     {
         return array(
-            'ALTER TABLE mytable RENAME INDEX idx_foo TO idx_bar',
+            'ALTER TABLE mytable RENAME INDEX `idx-foo` TO `idx-bar`',// Index name with special character in name (needs quotation on some platforms, e.g. Sqlite).
```
Should be `idx_foo` on the left side, no?

Ocramius commented on tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php:
```diff
@@ -57,7 +57,7 @@ protected function getQuotedAlterTableRenameIndexSQL()
     {
         return array(
-            'ALTER TABLE myschema.mytable RENAME INDEX idx_foo TO idx_bar',
+            'ALTER TABLE myschema.mytable RENAME INDEX `idx-foo` TO `idx-bar`',// Index name with special character in name (needs quotation on some platforms, e.g. Sqlite).
```
Should be `idx_foo` on the left side, no?

Ocramius commented on tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php:
```diff
@@ -78,7 +78,7 @@ protected function getQuotedAlterTableRenameIndexInSchemaSQL()
     {
         return array(
-            'ALTER TABLE mytable RENAME INDEX idx_foo TO idx_foo_renamed',
+            'ALTER TABLE mytable RENAME INDEX `idx-foo` TO `idx-foo-renamed`',// Index name with special character in name (needs quotation on some platforms, e.g. Sqlite).
```
Should be `idx_foo` on the left side, no?

gabrielrincon replied:
Ok, i will modify.

gabrielrincon replied:
@Ocramius   Today, i am usign doctrine and i have a field that name was datetime and the doctrine crash. the quota is necesary on the field name, same bug with a field named to.

Ocramius replied:
Have you seen http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/basic-mapping.html#quoting-reserved-words and http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/limitations-and-known-issues.html#identifier-quoting-and-legacy-databases ?
